server/utils/forms/inputs/addressInput.ts

```diff
@@ -0,0 +1,80 @@
+import { Request } from 'express'
+import * as ExpressValidator from 'express-validator'
+import { FormValidationResult } from '../formValidationResult'
+import { FormValidationError } from '../../formValidationError'
+import FormUtils from '../../formUtils'
+import PresenterUtils from '../../presenterUtils'
+
+export interface AddressErrorMessages {
+  addressLine1Empty: string
+  townOrCityEmpty: string
+  countyEmpty: string
+  postCodeEmpty: string
+  postCodeInvalid: string
+}
+
+export default class AddressInput {
+  constructor(
+    private readonly request: Request,
+    private readonly key: string,
+    private readonly errorMessages: AddressErrorMessages
+  ) {}
+
+  private readonly utils = new PresenterUtils(this.request.body)
+
+  private get keys() {
+    return {
+      addressLine1: `${this.key}-address-line-1`,
+      addressLine2: `${this.key}-address-line-2`,
+      townOrCity: `${this.key}-address-town-or-city`,
+      county: `${this.key}-address-county`,
+      postCode: `${this.key}-address-postcode`,
+    }
+  }
+
+  async validate(): Promise<FormValidationResult<string[]>> {
+    const validationResult = await FormUtils.runValidations({ request: this.request, validations: this.validations })
+    const error = this.extractErrors(validationResult)
+    if (error) {
+      return { value: null, error }
+    }
+
+    return {
+      value: this.address!,
+      error: null,
+    }
+  }
+
+  private get validations(): ExpressValidator.ValidationChain[] {
+    return [
+      ExpressValidator.body(this.keys.addressLine1)
+        .notEmpty({ ignore_whitespace: true })
+        .withMessage(this.errorMessages.addressLine1Empty),
+
+      ExpressValidator.body(this.keys.townOrCity)
+        .notEmpty({ ignore_whitespace: true })
+        .withMessage(this.errorMessages.townOrCityEmpty),
+
+      ExpressValidator.body(this.keys.county)
+        .notEmpty({ ignore_whitespace: true })
+        .withMessage(this.errorMessages.countyEmpty),
+
+      ExpressValidator.body(this.keys.postCode)
+        .notEmpty({ ignore_whitespace: true })
+        .withMessage(this.errorMessages.postCodeEmpty),
+      ExpressValidator.body(this.keys.postCode).isLength({ max: 8 }).withMessage(this.errorMessages.postCodeInvalid),
```

Comment:
could we apply some basic regex to this? 

i know it's a thorny subject but there's some examples here we could use https://stackoverflow.com/questions/164979/regex-for-matching-uk-postcodes

```
^([Gg][Ii][Rr] 0[Aa]{2})|((([A-Za-z][0-9]{1,2})|(([A-Za-z][A-Ha-hJ-Yj-y][0-9]{1,2})|(([AZa-z][0-9][A-Za-z])|([A-Za-z][A-Ha-hJ-Yj-y][0-9]?[A-Za-z]))))[0-9][A-Za-z]{2})$
```

Reply:
Sure. I can do that. Leigh indicated that validation on the fields is not important; but this seems like a worthwhile validation to add in.